xonsh/ptk_shell/shell.py

```diff
@@ -195,7 +203,21 @@ def singleline(
                 except (AttributeError, TypeError, ValueError):
                     print_exception()
 
+        if env["ENABLE_ASYNC_PROMPT"]:
+            # once the prompt is done, update it in background as each future is completed
+            updators = [
+                threading.Thread(target=self.update_prompt, args=(pt_name, futures))
```

Comment:
Do we really need more threads for handling each future? Aren't they already async? Also I don't see a locking mechanism down in `self.update_prompt` so this code seems to be prone to race conditions.

Reply:
I did try that. But we needed a thread to get result from those futures and update the prompt. There is no way to know that the prompt processing is done for those futures, (i.e. they can start calling prompt.invalidate)

Reply:
also I dont see any need for locking conditions. when the next set of futures come in and previous futures still running, it will get cancelled.

Reply:
My concern is this creating a new thread for each updater without any locking. Wouldn't that cause a race condition if multiple updaters try to update the prompt around the same time?

Reply:
calling invalidate is thread safe. The commented code, was there to do that, but I commented as it will not be needed, since the future gets cleared at every call, the threads listening on them will die off. Here we dont need locking, only to clear off the old updaters to free up resources.